### DISCUSSION TITLE
Rule Consolidation or skip Rule Class Creation

### DISCUSSION THREAD

**Discussion by EDKarlsson** [textX/textX#378] [Ideas]
Hi, first time posting here. I'm currently working on creating a pilot implementation of the SysML v2 specification using textX. The specialized EBNF grammar provided in the spec has a lot of additional rules that are used to simplify reading and parsing but if used 1 to 1 in the textX grammar creates a lot of additional elements that are unnecessary. These terminal rules are used within other rules so they are convenient to use to avoid clutter and improves readability.

This is an example of a feature type.
<img width="686" alt="Screen Shot 2022-03-27 at 10 59 47 AM" src="https://user-images.githubusercontent.com/7072370/160294355-2d3266ce-cb92-4e09-8acd-f032af2bf93c.png">

Would it be possible to add some kind of annotation to the grammar or keyword to allow a "roll-up" or consolidation of a rule into the parent rule? I am considering writing a preprocessor that translates rules with an annotation or keyword with the terminal rule.

I've finished the Root layer and am almost done with the Core layer of the Kernel Modeling Language.

I have other thoughts as well such as specifying the type of the class a rule represents without having to specify it in the grammar eg MyRule[type=SomeType] but I'm not sure about if it's needed at this point.
Also being able to specify a set of files to be loaded as builtin in types, eg: loading a model as a library to be used.

**Comment by goto40** [comment]
@EDKarlsson, It is really great that you started to implement SysML v2 with textx! I am looking forward to see some examples...

Regarding your question: You presented some possibilities to allow a type conversion/transformation/selection. I understand that it would be nice to include this in the grammar (as you mentioned correctly, it improves readability and results in a better documentation)...

If I understood it right, it is 
 * either a custom "annotation" or "tag" for a rule to be used in some user function/or model/rule processor- in that case we need to define the possible annotations somewhere (?),
 * or an explicit type conversion: this somehow sounds a little like the *actions* in Xtext (https://www.eclipse.org/Xtext/documentation/301_grammarlanguage.html)...  

A very simple "workaround" for the first solution would be to choose a certain naming of the rule (e.g., MyRuleWithTagABC) and use the rule name (name ends with "TagABC" ) as "annotation" (not very elegant, but easy to do).

@igordejanovic, what do you think?

**Reply by EDKarlsson** [reply]
I think using a rule name would be an acceptable solution which I assumes avoid having to update the actual structure of the textX grammar. I would also be fine to have a parameter that to “roll-up” any rule that doesn’t have an explicit class defined. 

In the case of SysML 2 there will be a lot of object and model processing so I am creating an explicit class for each element in the language.

**Comment by EDKarlsson** [comment]
There is also a situation I am dealing with the SysML 2 spec where there are types such as `Membership` which isn't defined in the grammar but is a model element that exists in the language. It would be nice to be able to type a rule as that element but not have to have a grammar rule to force that type. I have the classes for the rule inherit from the `Membership` element which is the best I think I can do right now.

I could possibly just create a Membership rule that is never referenced but I've noticed if I use type=[...] it wants to reference the rule and it's statements.

**Comment by EDKarlsson** [comment]
Here is an example of a simple KerML model which shows the additional elements that are created. In a small model this isn't a problem but when models get larger this will become exponential.

![elements dot](https://user-images.githubusercontent.com/7072370/160300550-94f74483-f55e-4b8d-855b-81239d07090b.png)

**Comment by goto40** [comment]
Did you already have a look at https://github.com/Systems-Modeling/SysML-v2-Pilot-Implementation?

Maybe you find some inspiration how they solved some similar problems... I think it would be a nice thing to have a light-weight python library to read sysml models...

For your example: Maybe it would ease the discussion if you provide (1) the model-text with the (2) grammar you employed for the image above - and how you think (3) the grammar should ideally look like (from the sys-ml spec.) together with the (4) "extra-elements" created in your implementation (which you want to get rid of). If the grammar is too large just show a snippet - I assume the screenshot in your first comment is probably such a snippet (from the sys-ml spec).

**Reply by EDKarlsson** [reply]
Unfortunately, I'm waiting on approval from management to post my project repository on our org Github repo (https://github.com/open-mbee), so I am not sure what pieces of the code I'm allowed to share at the moment. Even though this is a side project, it is related to the work I'm doing and want to avoid getting into trouble. That being said, I am very anxious to share it and I think it's silly that I can't just post it to Github.

The Pilot implementation is a good example but heavily relies on Ecore/EMF, Eclipse plugins, etc.. which is the driver for me to implement it using textX. I've talked to Ed Seidewitz (primary developer of SysML v2 + implementation) and the only way they're able to get the language to work properly in their environment is to use a deprecated feature of Xtext wherein they can create rule fragments.

Yes, the screenshot is an output from my project. I'll type up a separate spec of what I am thinking.

**Comment by igordejanovic** [comment]
Hi @EDKarlsson. It's great to hear that you are working on pilot implementation of SysML v2 in textX! I'm interested  in SysML v2 also.
Is it going to be an open-source project?

There is (a very old) issue] #13 that suggest "include rules". It looks to me that it would be a solution to the problem of removing unnecessary clutter in the grammar without creating additional objects in the model.  

Type conversion is also something that might be interesting to implement. I think we had a discussion in the past about type conversion (especially match rules) but can't seem to find it at the moment.

I think it would be best to split each requirement in a separate issue for easier tracking.

As @goto40 suggested, if you can (i.e. your work is not proprietary) you could share a bit more details so that we have some material to base our work and tests on.

**Reply by EDKarlsson** [reply]
I am planning on open-sourcing the project and will most likely put it on https://github.com/open-mbee. I work on tools for Systems Engineering/MBSE.

In my opinion, the work isn't proprietary but I need to follow my organization's guidelines for submitting to open-source. I will gladly share and open it up for community development since this is mostly a hobby implementation.

To be honest, it started a project to prove that some vendors are complaining more than necessary about implementing the SysML v2 spec.

**Reply by igordejanovic** [reply]
It's nice to hear where textX is used. I'm looking forward to see your implementation online.

**Comment by igordejanovic** [comment]
> Also being able to specify a set of files to be loaded as builtin in types, eg: loading a model as a library to be used.

You can check [BuiltIn models feature](http://textx.github.io/textX/latest/scoping/#builtin-models). It is used e.g. [in pyFlies](https://github.com/pyflies/pyflies/blob/main/pyflies/lang/__init__.py#L29). Is that what you have been thinking?

**Reply by EDKarlsson** [reply]
Oh that's perfect! Yes that is what I am looking for. I won't sugar coat it but with how my mind works I find the documentation hard to navigate the documentation at times.

**Reply by igordejanovic** [reply]
We are always looking for ways to improve the docs. There is a search bar in the upper left corner so it  might help. If you have suggestions on the structure/navigation improvement do not hesitate to post an issue. We treat problems with the docs as bugs.

**Reply by EDKarlsson** [reply]
Sounds good, I will make notes on how I navigate the docs when looking for topics.

**Comment by EDKarlsson** [comment]
I think the simplest example (using sysml 2 grammar) I can write of what I'm interested in is the Identification rule.

```
Identification (e : Element, m : Membership) =
    ( '<' humanId = Name '>' )? ( m.memberName = NAME )?

Element (m : Membership) : Element =
    'element' Identification(this, m) ElementBody(this)
```

`Identification(...)` is used everywhere and in the sysml2 grammar takes 2 parameters. The subject of the rule (this) and the membership property of the calling element. Taking a simple approach to this, which will need to be refactored later or addressed through processing is to write 

```
Element (m : Membership) : Element =
    'element' (  '<' humanId = ID '>' name=ID | '<' name=ID '>' | name=ID )?
    ElementBody(this)
```

or properly with sysml2 attribute names

```
Element (m : Membership) : Element =
    'element' (  '<' humanId = ID '>' memberName=ID | '<' memberName=ID '>' | memberName=ID )?
    ElementBody(this)
```

Per the SysML spec it isn't technically correct because the unique identifier attribute of an element, which is a UUID, is called `identifier` which is what should be unique in the model. However, in textX the `name` attribute is the unique property.
This isn't a real big problem because I can handle this in processing.

What is exhausting is having to replace all the instances where `Identification(this, m)` is called within the grammar. Also, there is no easy way of passing the `name` attribute back to the calling rule which allows for referencing and keeping elements unique.

**Reply by EDKarlsson** [reply]
Thinking about it, a simple C macro preprocessor on the grammar would suffice. eg
```
#Identification ( '<' humanId = ID '>' name=ID | '<' name=ID '>' | name=ID )?
```

**Reply by igordejanovic** [reply]
Yes, if it is a matter of fragment replacement then using a preprocessor might be a way to go as a workaround. 

> However, in textX the name attribute is the unique property.
> This isn't a real big problem because I can handle this in processing.

`name` attribute is used for default reference resolving but technically you can override [scope providers](http://textx.github.io/textX/latest/scoping/#textx-scoping) which are in charge of reference resolving. Thus, you can implement your own strategy of naming and reference resolving. You can probably reuse already existing scope providers for that. @goto40 might be able to provide more details.

**Reply by EDKarlsson** [reply]
Ah okay, my intuition was correct on how to approach it. I always try to go for the simplest solution first then refactor. Live by Knuth's Premature optimization is the root of all evil.

**Reply by igordejanovic** [reply]
Anyway, if you have ideas for #13, please add there your thoughts on what you think would be a nice solution from the user perspective. It seems that some kind of parametrization would be nice to support your use-cases.
